src/protocols/rdp/argv.c

```diff
@@ -0,0 +1,192 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include "config.h"
+#include "argv.h"
+#include "rdp.h"
+
+#include <guacamole/protocol.h>
+#include <guacamole/socket.h>
+#include <guacamole/user.h>
+
+#include <pthread.h>
+#include <stdlib.h>
+#include <string.h>
+
+/**
+ * All RDP connection settings which may be updated by unprivileged users
+ * through "argv" streams.
+ */
+typedef enum guac_rdp_argv_setting {
+
+    /**
+     * The username of the connection.
+     */
+    GUAC_RDP_ARGV_SETTING_USERNAME,
+            
+    /**
+     * The password to authenticate the connection.
+     */
+    GUAC_RDP_ARGV_SETTING_PASSWORD,
+            
+    /**
+     * The domain to use for connection authentication.
+     */
+    GUAC_RDP_ARGV_SETTING_DOMAIN,
+    
+} guac_rdp_argv_setting;
+
+/**
+ * The value or current status of a connection parameter received over an
+ * "argv" stream.
+ */
+typedef struct guac_rdp_argv {
+
+    /**
+     * The specific setting being updated.
+     */
+    guac_rdp_argv_setting setting;
+
+    /**
+     * Buffer space for containing the received argument value.
+     */
+    char buffer[GUAC_RDP_ARGV_MAX_LENGTH];
+
+    /**
+     * The number of bytes received so far.
+     */
+    int length;
+
+} guac_rdp_argv;
+
+/**
+ * Handler for "blob" instructions which appends the data from received blobs
+ * to the end of the in-progress argument value buffer.
+ *
+ * @see guac_user_blob_handler
+ */
+static int guac_rdp_argv_blob_handler(guac_user* user,
+        guac_stream* stream, void* data, int length) {
+
+    guac_rdp_argv* argv = (guac_rdp_argv*) stream->data;
+    
+    /* Calculate buffer size remaining, including space for null terminator,
+     * adjusting received length accordingly */
+    int remaining = sizeof(argv->buffer) - argv->length - 1;
+    if (length > remaining)
+        length = remaining;
+
+    /* Append received data to end of buffer */
+    memcpy(argv->buffer + argv->length, data, length);
+    argv->length += length;
+
+    return 0;
+
+}
+
+/**
+ * Handler for "end" instructions which applies the changes specified by the
+ * argument value buffer associated with the stream.
+ *
+ * @see guac_user_end_handler
+ */
+static int guac_rdp_argv_end_handler(guac_user* user,
+        guac_stream* stream) {
+
+    guac_client* client = user->client;
+    guac_rdp_client* rdp_client = (guac_rdp_client*) client->data;
+    guac_rdp_settings* settings = rdp_client->settings;
+    
+    /* Append null terminator to value */
+    guac_rdp_argv* argv = (guac_rdp_argv*) stream->data;
+    argv->buffer[argv->length] = '\0';
```

Comment:
Is there a guarantee that there will be space for this null terminator within `argv->buffer`?

Reply:
I haven't looked, but my assumption (yes, I know) is that when `argv->length` is calculated, it cannot be higher than the length of the buffer, so the worst that would happen is this kills the last character of the provided string?

Also, wasn't sure if you were saying in the other comment that using `strndup()` makes this unnecessary??

Reply:
If `buffer` points to a buffer of exactly `length` bytes, then `buffer[length]` is outside the buffer, not the last byte of the buffer.

Reply:
Ah, yes, that makes sense.  Is adding the NULL-terminator necessary?  I'm guessing I copied it from other code - probably from the code that updates terminal values SSH/Telnet sessions.  I'm happy to remove it completely if there's no need for it.

Reply:
Yep, definitely necessary. Nothing handling the received data as a string will be able to function safely without that null terminator.

The beauty of `strndup(foo, length)` is that it will automatically add that null terminator and ensure that there is space for it within the newly-allocated buffer. Assuming the string within the buffer is indeed `length` characters long, `strndup(..., length)` would allocate buffer space of at least `length + 1` bytes.

Reply:
Okay, so do the changes I've made address this, since I've moved it to `strndup()`, or does the null terminator need to be added to the original `argv->buffer` string?

Reply:
From re-reading your above comments, adding it manually is not required when using `strndup()`